File: Packages/OuraProtocol/Sources/OuraProtocol/OuraEvents.swift
import Foundation

// OuraEvents: the decoded value structs the driver emits (OURA_PROTOCOL.md s6). Each carries the
// record's ringTimestamp (the ring-clock value; the app anchors it to UTC via the 0x42 time-sync /
// 0x85 RTC events) plus the decoded signal. Pure value types, no CoreBluetooth.
//
// Per-sample timestamps inside a record (IBI/temp/HRV/SpO2) walk backward from the event time by each
// sample's own duration (OURA_PROTOCOL.md s6); to stay platform-pure and avoid baking a clock model
// into the decoders, the structs carry the raw ring/sample offsets and let the app's mapping layer
// (OuraStreamMapping) apply the anchor. Honest-data invariant: a short/malformed record decodes to
// nil upstream, so these structs only ever hold real decoded values.

/// One decoded inter-beat interval (and optional amplitude), in milliseconds.
public struct OuraIBI: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let ibiMs: Int
    public let amplitude: Int?
    public init(ringTimestamp: UInt32, ibiMs: Int, amplitude: Int? = nil) {
        self.ringTimestamp = ringTimestamp; self.ibiMs = ibiMs; self.amplitude = amplitude
    }
}

/// One decoded heart-rate value in BPM (derived from a live-HR push IBI, OURA_PROTOCOL.md s5.6).
public struct OuraHR: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let bpm: Int
    public let ibiMs: Int
    public init(ringTimestamp: UInt32, bpm: Int, ibiMs: Int) {
        self.ringTimestamp = ringTimestamp; self.bpm = bpm; self.ibiMs = ibiMs
    }
}

/// One decoded HRV (RMSSD-derived) sample from the ring's own 0x5D tag (OURA_PROTOCOL.md s6.9).
/// NOOP also reconstructs RMSSD itself from the IBI streams for its own scoring; this is the ring's
/// open HRV tag, NOT Oura's encrypted readiness score.
public struct OuraHRV: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let timeMs: Int
    public let b1: Int
    public let b2: Int
    public init(ringTimestamp: UInt32, timeMs: Int, b1: Int, b2: Int) {
        self.ringTimestamp = ringTimestamp; self.timeMs = timeMs; self.b1 = b1; self.b2 = b2
    }
}

/// One decoded SpO2 sample. `value` is the raw SpO2 reading; `unit` documents its scale.
public struct OuraSpO2: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let value: Int
    public let unit: String
    public init(ringTimestamp: UInt32, value: Int, unit: String = "raw") {
        self.ringTimestamp = ringTimestamp; self.value = value; self.unit = unit
    }
}

/// One decoded skin-temperature sample in hundredths of a degree C scaled to C (value already / 100).
public struct OuraTemp: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let celsius: Double
    public init(ringTimestamp: UInt32, celsius: Double) {
        self.ringTimestamp = ringTimestamp; self.celsius = celsius
    }
}

/// One decoded battery reading (OURA_PROTOCOL.md s6.10). `percent` is read at body[0]; `voltageMv`
/// is the [4..6] fallback estimate (fixture-validated per generation, may be nil).
public struct OuraBattery: Equatable, Sendable, Codable {
    public let percent: Int
    public let voltageMv: Int?
    public let charging: Bool?
    public init(percent: Int, voltageMv: Int? = nil, charging: Bool? = nil) {
        self.percent = percent; self.voltageMv = voltageMv; self.charging = charging
    }
}

/// Sleep phase code (OURA_PROTOCOL.md s6.12): 2-bit codes 0=awake, 1=light, 2=deep, 3=REM.
public enum OuraSleepStage: Int, Sendable, Equatable, Codable {
    case awake = 0
    case light = 1
    case deep = 2
    case rem = 3
}

/// One decoded sleep-phase code in order within a 0x4E/0x5A record (OURA_PROTOCOL.md s6.12).
public struct OuraSleepPhase: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let index: Int          // position within the record's phase sequence
    public let stage: OuraSleepStage
    public init(ringTimestamp: UInt32, index: Int, stage: OuraSleepStage) {
        self.ringTimestamp = ringTimestamp; self.index = index; self.stage = stage
    }
}

/// Motion state (OURA_PROTOCOL.md s6.13): 0 NO_MOTION, 1 RESTLESS, 2 TOSSING, 3 ACTIVE.
public enum OuraMotionState: Int, Sendable, Equatable, Codable {
    case noMotion = 0
    case restless = 1
    case tossing = 2
    case active = 3
}

/// One decoded motion-state code from a 0x6B motion_period record (OURA_PROTOCOL.md s6.13).
public struct OuraMotion: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let index: Int
    public let state: OuraMotionState
    public init(ringTimestamp: UInt32, index: Int, state: OuraMotionState) {
        self.ringTimestamp = ringTimestamp; self.index = index; self.state = state
    }
}

/// Device lifecycle state (OURA_PROTOCOL.md s6.15) decoded from a 0x45/0x53 record.
public struct OuraState: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let stateCode: Int
    public let text: String?
    public init(ringTimestamp: UInt32, stateCode: Int, text: String? = nil) {
        self.ringTimestamp = ringTimestamp; self.stateCode = stateCode; self.text = text
    }
}

/// A UTC anchor / time-sync event (OURA_PROTOCOL.md s6.11): epoch ms + timezone offset seconds.
public struct OuraTimeSync: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let epochMs: Int64
    public let tzOffsetSeconds: Int
    public init(ringTimestamp: UInt32, epochMs: Int64, tzOffsetSeconds: Int) {
        self.ringTimestamp = ringTimestamp; self.epochMs = epochMs; self.tzOffsetSeconds = tzOffsetSeconds
    }
}

/// A secondary 1-second-granularity RTC beacon (OURA_PROTOCOL.md s6.15, tag 0x85).
public struct OuraRtcBeacon: Equatable, Sendable, Codable {
    public let ringTimestamp: UInt32
    public let unixSeconds: Int
    public init(ringTimestamp: UInt32, unixSeconds: Int) {
        self.ringTimestamp = ringTimestamp; self.unixSeconds = unixSeconds
    }
}

// MARK: - Tier-B (UNVERIFIED) decoded events

/// A Tier-B sleep summary value (OURA_PROTOCOL.md s6.12). UNVERIFIED layout; carries the raw payload
/// bytes plus the tag so a fixture test can validate before scoring trusts it. The driver only emits
/// this when allowTierB is set, and it is never folded into scoring silently.
public struct OuraTierBSummary: Equatable, Sendable, Codable {
    public let tag: UInt8
    public let ringTimestamp: UInt32
    public let rawPayload: [UInt8]
    public let kind: String        // "sleep_summary" / "activity" / "real_steps" / "spo2_smoothed"
    public init(tag: UInt8, ringTimestamp: UInt32, rawPayload: [UInt8], kind: String) {
        self.tag = tag; self.ringTimestamp = ringTimestamp; self.rawPayload = rawPayload; self.kind = kind
    }
}

// MARK: - The emitted event union

/// What OuraDriver.ingest(record:) emits. A single record can yield several events (e.g. an IBI+amp
/// record carries up to 6 IBIs). Tier-B events are wrapped in .tierB and only emitted when the driver
/// is configured to allow them; they must never feed scoring without passing a real-capture fixture.
public enum OuraEvent: Equatable, Sendable {
    case hr(OuraHR)
    case ibi(OuraIBI)
    case hrv(OuraHRV)
    case spo2(OuraSpO2)
    case temp(OuraTemp)
    case battery(OuraBattery)
    case sleepPhase(OuraSleepPhase)
    case motion(OuraMotion)
    case state(OuraState)
    case timeSync(OuraTimeSync)
    case rtcBeacon(OuraRtcBeacon)
    case debugText(ringTimestamp: UInt32, text: String)
    /// A Tier-B (UNVERIFIED) decoded value. Gated behind OuraDriver.allowTierB. Per the brief's TIER
    /// DISCIPLINE: do not let Tier B feed values silently.
    case tierB(OuraTierBSummary)

    /// True for Tier-B events, so a consumer can assert none leaked into a Tier-A-only sink.
    public var isTierB: Bool {
        if case .tierB = self { return true }
        return false
    }
}

File: android/app/src/main/java/com/noop/oura/OuraEvents.kt
package com.noop.oura

// OuraEvents: the decoded value structs the driver emits (OURA_PROTOCOL.md s6). Kotlin twin of
// OuraEvents.swift. Each carries the record's ringTimestamp (the ring-clock value; the app anchors it
// to UTC via the 0x42 time-sync / 0x85 RTC events) plus the decoded signal. Pure value types, no
// android.bluetooth.
//
// DIVERGENCE FROM SWIFT: Swift's UInt32 ringTimestamp becomes a Long holding the unsigned 32-bit
// value (0..0xFFFFFFFF), and Swift's Int64 epoch becomes Long. Values and layouts are identical.
//
// Per-sample timestamps inside a record (IBI/temp/HRV/SpO2) walk backward from the event time by each
// sample's own duration (OURA_PROTOCOL.md s6); to stay platform-pure and avoid baking a clock model
// into the decoders, the structs carry the raw ring/sample offsets and let the app's mapping layer
// apply the anchor. Honest-data invariant: a short/malformed record decodes to null upstream, so
// these structs only ever hold real decoded values.

/** One decoded inter-beat interval (and optional amplitude), in milliseconds. */
data class OuraIBI(val ringTimestamp: Long, val ibiMs: Int, val amplitude: Int? = null)

/** One decoded heart-rate value in BPM (derived from a live-HR push IBI, OURA_PROTOCOL.md s5.6). */
data class OuraHR(val ringTimestamp: Long, val bpm: Int, val ibiMs: Int)

/**
 * One decoded HRV (RMSSD-derived) sample from the ring's own 0x5D tag (OURA_PROTOCOL.md s6.9).
 * NOOP also reconstructs RMSSD itself from the IBI streams for its own scoring; this is the ring's
 * open HRV tag, NOT Oura's encrypted readiness score.
 */
data class OuraHRV(val ringTimestamp: Long, val timeMs: Int, val b1: Int, val b2: Int)

/** One decoded SpO2 sample. `value` is the raw SpO2 reading; `unit` documents its scale. */
data class OuraSpO2(val ringTimestamp: Long, val value: Int, val unit: String = "raw")

/** One decoded skin-temperature sample in degrees C (value already / 100). */
data class OuraTemp(val ringTimestamp: Long, val celsius: Double)

/**
 * One decoded battery reading (OURA_PROTOCOL.md s6.10). `percent` is read at body[0]; `voltageMv`
 * is the [4..6] fallback estimate (fixture-validated per generation, may be null).
 */
data class OuraBattery(val percent: Int, val voltageMv: Int? = null, val charging: Boolean? = null)

/** Sleep phase code (OURA_PROTOCOL.md s6.12): 2-bit codes 0=awake, 1=light, 2=deep, 3=REM. */
enum class OuraSleepStage(val raw: Int) {
    AWAKE(0),
    LIGHT(1),
    DEEP(2),
    REM(3);

    companion object {
        private val byRaw = entries.associateBy { it.raw }
        fun fromRaw(raw: Int): OuraSleepStage? = byRaw[raw]
    }
}

/** One decoded sleep-phase code in order within a 0x4E/0x5A record (OURA_PROTOCOL.md s6.12). */
data class OuraSleepPhase(val ringTimestamp: Long, val index: Int, val stage: OuraSleepStage)

/** Motion state (OURA_PROTOCOL.md s6.13): 0 NO_MOTION, 1 RESTLESS, 2 TOSSING, 3 ACTIVE. */
enum class OuraMotionState(val raw: Int) {
    NO_MOTION(0),
    RESTLESS(1),
    TOSSING(2),
    ACTIVE(3);

    companion object {
        private val byRaw = entries.associateBy { it.raw }
        fun fromRaw(raw: Int): OuraMotionState? = byRaw[raw]
    }
}

/** One decoded motion-state code from a 0x6B motion_period record (OURA_PROTOCOL.md s6.13). */
data class OuraMotion(val ringTimestamp: Long, val index: Int, val state: OuraMotionState)

/** Device lifecycle state (OURA_PROTOCOL.md s6.15) decoded from a 0x45/0x53 record. */
data class OuraState(val ringTimestamp: Long, val stateCode: Int, val text: String? = null)

/** A UTC anchor / time-sync event (OURA_PROTOCOL.md s6.11): epoch ms + timezone offset seconds. */
data class OuraTimeSync(val ringTimestamp: Long, val epochMs: Long, val tzOffsetSeconds: Int)

/** A secondary 1-second-granularity RTC beacon (OURA_PROTOCOL.md s6.15, tag 0x85). */
data class OuraRtcBeacon(val ringTimestamp: Long, val unixSeconds: Long)

// MARK: - Tier-B (UNVERIFIED) decoded events

/**
 * A Tier-B sleep summary value (OURA_PROTOCOL.md s6.12). UNVERIFIED layout; carries the raw payload
 * bytes plus the tag so a fixture test can validate before scoring trusts it. The driver only emits
 * this when allowTierB is set, and it is never folded into scoring silently.
 */
data class OuraTierBSummary(
    val tag: Int,
    val ringTimestamp: Long,
    val rawPayload: IntArray,
    val kind: String,          // "sleep_summary" / "activity" / "real_steps" / "spo2_smoothed"
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is OuraTierBSummary) return false
        return tag == other.tag && ringTimestamp == other.ringTimestamp &&
            rawPayload.contentEquals(other.rawPayload) && kind == other.kind
    }

    override fun hashCode(): Int {
        var h = tag
        h = 31 * h + ringTimestamp.hashCode()
        h = 31 * h + rawPayload.contentHashCode()
        h = 31 * h + kind.hashCode()
        return h
    }
}

// MARK: - The emitted event union

/**
 * What OuraDriver.ingest(record:) emits. A single record can yield several events (e.g. an IBI+amp
 * record carries up to 6 IBIs). Tier-B events are wrapped in TierB and only emitted when the driver
 * is configured to allow them; they must never feed scoring without passing a real-capture fixture.
 *
 * Kotlin twin of the Swift `OuraEvent` enum-with-associated-values, modelled as a sealed class.
 */
sealed class OuraEvent {
    data class Hr(val value: OuraHR) : OuraEvent()
    data class Ibi(val value: OuraIBI) : OuraEvent()
    data class Hrv(val value: OuraHRV) : OuraEvent()
    data class Spo2(val value: OuraSpO2) : OuraEvent()
    data class Temp(val value: OuraTemp) : OuraEvent()
    data class Battery(val value: OuraBattery) : OuraEvent()
    data class SleepPhaseEvent(val value: OuraSleepPhase) : OuraEvent()
    data class MotionEvent(val value: OuraMotion) : OuraEvent()
    data class StateEvent(val value: OuraState) : OuraEvent()
    data class TimeSyncEvent(val value: OuraTimeSync) : OuraEvent()
    data class RtcBeaconEvent(val value: OuraRtcBeacon) : OuraEvent()
    data class DebugTextEvent(val ringTimestamp: Long, val text: String) : OuraEvent()

    /**
     * A Tier-B (UNVERIFIED) decoded value. Gated behind OuraDriver.allowTierB. Per the brief's TIER
     * DISCIPLINE: do not let Tier B feed values silently.
     */
    data class TierB(val value: OuraTierBSummary) : OuraEvent()

    /** True for Tier-B events, so a consumer can assert none leaked into a Tier-A-only sink. */
    val isTierB: Boolean get() = this is TierB
}
